[Attempt 2] refactor: Clean up lms/envs/production.py cruft

kdmccormick opened the pull request:
## Description

This PR has 3 commits:

1. We re-introduce https://github.com/openedx/edx-platform/pull/36115, which was reverted by https://github.com/openedx/edx-platform/pull/36129 due to a typo.
2. We fix said typo
3. We tighten up our backwards-compatibility of `ENV_TOKENS` and a few other "auxillary" settings dictionaries, out of an abundance of caution and a desire not to break any existing Django plugins unintentionally. See commit description for more info.

Altogether, this PR makes exactly ONE change to lms/envs/production.py, and that is the removal of `KEYS_WITH_MERGED_VALUES` from the settings namespace. This "setting" was an internal implementation detail of production.py, which should have been prefixed with an underscore. I am not considering this a breaking change.

Otherwise, this PR should exactly preserve the global namespace of lms/envs/production.py given any LMS_CFG yaml file.

## Testing Instructions

Here is a little script you can paste into edx-platform/diff_settings.sh (note the Tutor vs non-Tutor difference):

```bash
#!/usr/bin/env bash
# Usage: ./diff_settings.sh GIT_REF_A GIT_REF_B
# Will fail if git state is not clean.

set -xeuo pipefail  # Be verbose and strict

REF_A=$1
REF_B=$2
DIR_A=dump_settings_a
DIR_B=dump_settings_b

SETTINGS=( envs.devstack envs.production )  #  If not using Tutor
SETTINGS=( envs.tutor.production envs.devstack envs.tutor.development )  # If using Tutor

rm -rf "$DIR_A" "$DIR_B"
mkdir "$DIR_A" "$DIR_B"

for system in cms lms ; do
    for settings in "${SETTINGS[@]}" ; do
        module="${system}.${settings}"
        git checkout "$REF_A"
        DJANGO_SETTINGS_MODULE="$module" ./manage.py "$system" dump_settings > "$DIR_A/$module.json"
        git checkout "$REF_B"
        DJANGO_SETTINGS_MODULE="$module" ./manage.py "$system" dump_settings > "$DIR_B/$module.json"
    done
done

diff "$DIR_A" "$DIR_B" && echo "No difference!"
```

Ensure you have this PR's branch checked out and a updated master branch, and then run:
```bash
./diff_settings.sh master kdmccormick/lms-envs-production-cleanup-2
```
You should see that the only diff is the removal of `KEYS_WITH_MERGED_VALUES` from the LMS files:

```
+ diff dump_settings_a dump_settings_b
diff dump_settings_a/lms.envs.devstack.json dump_settings_b/lms.envs.devstack.json
2770,2780d2769
<     "KEYS_WITH_MERGED_VALUES": [
<         "FEATURES",
<         "TRACKING_BACKENDS",
<         "EVENT_TRACKING_BACKENDS",
<         "JWT_AUTH",
<         "CELERY_QUEUES",
<         "MKTG_URL_LINK_MAP",
<         "MKTG_URL_OVERRIDES",
<         "REST_FRAMEWORK",
<         "EVENT_BUS_PRODUCER_CONFIG"
<     ],
diff dump_settings_a/lms.envs.tutor.development.json dump_settings_b/lms.envs.tutor.development.json
2709,2719d2708
<     "KEYS_WITH_MERGED_VALUES": [
<         "FEATURES",
<         "TRACKING_BACKENDS",
<         "EVENT_TRACKING_BACKENDS",
<         "JWT_AUTH",
<         "CELERY_QUEUES",
<         "MKTG_URL_LINK_MAP",
<         "MKTG_URL_OVERRIDES",
<         "REST_FRAMEWORK",
<         "EVENT_BUS_PRODUCER_CONFIG"
<     ],
diff dump_settings_a/lms.envs.tutor.production.json dump_settings_b/lms.envs.tutor.production.json
2446,2456d2445
<     "KEYS_WITH_MERGED_VALUES": [
<         "FEATURES",
<         "TRACKING_BACKENDS",
<         "EVENT_TRACKING_BACKENDS",
<         "JWT_AUTH",
<         "CELERY_QUEUES",
<         "MKTG_URL_LINK_MAP",
<         "MKTG_URL_OVERRIDES",
<         "REST_FRAMEWORK",
<         "EVENT_BUS_PRODUCER_CONFIG"
<     ],
```

kdmccormick committed:
revert: revert: refactor: Clean up lms/envs/production.py cruft

This reintroduces commit 15939232d547b751a06e9cb0c1c39ba78144bade,
which was reverted due to a typo.

The typo is fixed in the commit immediately following this one.

Co-Authored-By: Feanil Patel <feanil@axim.org>

kdmccormick committed:
fix: SHARED_kCOOKIE_DOMAIN -> SHARED_COOKIE_DOMAIN

robrap commented:
Note: Test plan being discussed in Slack, with request not to merge without a plan.

feanil reviewed (commented):
I took a look through again with a finer view of the additions to make sure there weren't any typos and I was able to bring up the LMS and CMS from this version of the settings file and the minimal.yaml locally so I think it's good to go but I'll wait till we add the unit test that we were discussing in slack before providing the final approval.

kdmccormick commented:
(rephrasing from Slack)  I don't see any way to unit test this PR, since unit tests use lms/envs/test.py, which is based directly on lms/envs/common.py, not lms/envs/production.py. I think we will have to rely on code review, manual testing, and the PR sandbox for verification.

open-craft-grove commented:
Sandbox deployment failed 💥
Please check the settings and requirements.
Retry deployment by pushing a new commit or updating the requirements/settings in the pull request's description.
📜 [Failure Logs](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36131-7daceb-51564656-8935654316.log)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36131-7daceb-51564656-8935654316-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36131-7daceb-51564656-8935654316-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36131-7daceb-51564656-8935654316-tutor-requirements.txt)

open-craft-grove commented:
Sandbox deployment successful 🚀
🎓 [LMS](https://pr-36131-7daceb.sandboxes.opencraft.hosting)
📝 [Studio](https://studio.pr-36131-7daceb.sandboxes.opencraft.hosting)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36131-7daceb-51564656-8935654297-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36131-7daceb-51564656-8935654297-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36131-7daceb-51564656-8935654297-tutor-requirements.txt)

kdmccormick commented:
Converting this to draft until I can manually reverify it using `dump_settings` from https://github.com/openedx/edx-platform/pull/36162

open-craft-grove commented:
Sandbox deployment failed 💥
Please check the settings and requirements.
Retry deployment by pushing a new commit or updating the requirements/settings in the pull request's description.
📜 [Failure Logs](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36131-7daceb-51564656-8964578310.log)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36131-7daceb-51564656-8964578310-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36131-7daceb-51564656-8964578310-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36131-7daceb-51564656-8964578310-tutor-requirements.txt)

kdmccormick committed:
fix: Exactly preserve legacy settings dicts; rm KEYS_WITH_MERGED_VALUES

In the near term, we wish to precisely preserve the existing values of
all Django settings exposed by lms/envs/production.py in order to avoid
breaking legacy Django plugins without a proper announcement.

That includes preserving the behavior of these old, redundant dicts:
* ENV_TOKENS
* AUTH_TOKENS
* ENV_FEATURES
* ENV_CELERY_QUEUES
* ALTERNATE_QUEUE_ENVS

Particularly, it means we need to ensure that updates to Django
settings are reflected in these dicts. The most reliable way to do that
is to change the yaml-loading logic so that these values are aliased
to the corresponding values in the global namespace rather than
deep-copied.

Finally, we remove KEYS_WITH_MERGED_VALUES from the global namespace,
and inline the remaining list. We have modified the list (specifically,
we dropped the no-op MKTG_URL_OVERRIDES). Plugins should not be counting
on the value of the list, so we remove it.

open-craft-grove commented:
Sandbox deployment successful 🚀
🎓 [LMS](https://pr-36131-7daceb.sandboxes.opencraft.hosting)
📝 [Studio](https://studio.pr-36131-7daceb.sandboxes.opencraft.hosting)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36131-7daceb-51564656-8977568161-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36131-7daceb-51564656-8977568161-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36131-7daceb-51564656-8977568161-tutor-requirements.txt)

kdmccormick commented:
@robrap @feanil This is ready for re-review.

open-craft-grove commented:
Sandbox deployment successful 🚀
🎓 [LMS](https://pr-36131-7daceb.sandboxes.opencraft.hosting)
📝 [Studio](https://studio.pr-36131-7daceb.sandboxes.opencraft.hosting)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36131-7daceb-51564656-8979929921-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36131-7daceb-51564656-8979929921-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36131-7daceb-51564656-8979929921-tutor-requirements.txt)

open-craft-grove commented:
Sandbox deployment successful 🚀
🎓 [LMS](https://pr-36131-7daceb.sandboxes.opencraft.hosting)
📝 [Studio](https://studio.pr-36131-7daceb.sandboxes.opencraft.hosting)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36131-7daceb-51564656-8987245850-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36131-7daceb-51564656-8987245850-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36131-7daceb-51564656-8987245850-tutor-requirements.txt)

robrap commented:
Thanks @kdmccormick. I'm going to have @dianakhuang take over anything required to get this over-the-line from 2U's end. Thanks!

kdmccormick commented:
Sounds good @robrap . Since non-SRE engineers cannot run management commands on stage (unless that changed since I left), I would recommend using a prod-like yaml file to compare the output of `dump_settings` for this PR vs for master before I merge, and ensure that `KEYS_WITH_MERGED_VALUES` is the only differing setting.

robrap commented:
Yes. @dianakhuang is working on getting a prod-like yaml file on a sandbox.

kdmccormick commented:
Ah sorry @robrap , I misread your message, I thought you were taking over for Diana, not the other way around. We're all extra aligned, now 😆 

@dianakhuang , let me know if there's anything I can clarify on dump_settings or this PR.

feanil commented:
FYI, I was able to run with this locally and bring up the LMS and CMS so no issues with the basic settings being set from here or in minimal.yml.

dianakhuang commented:
I'm working on getting pre and post dumps of the settings to diff them, and once I have gone over them, I can approve.

dianakhuang commented:
Okay, the only diffs I saw in the sandbox version were `KEYS_WITH_MERGED_VALUES` and the revision differences. I am 🤞  that this means that we're good to go from our side.
